sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1854,3 +1855,37 @@ object OptimizeLimitZero extends Rule[LogicalPlan] {
       empty(ll)
   }
 }
+
+/**
+ * Infers filters from [[Generate]], such that rows that would have been removed
+ * by this [[Generate]] can be removed earlier - before joins and in data sources.
+ */
+object InferFiltersFromGenerate extends Rule[LogicalPlan] {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+    case generate @ Generate(e: ExplodeBase, _, false, _, _, _)
+      if e.deterministic && !e.child.foldable =>
```

Comment:
We need this foldable check?

Reply:
We would get a foldable filter from this. 
99% of times it would be like `size([1, 2, 3]) > 0 && isNotNull([1, 2, 3])`.
But now that you mention it, yes it would be handled by `ConstantFolding` and in rare occasions it might be replaced with a false literal. 

I don't have a strong opinion on this.

Reply:
Ah, I see. okay, let's just check `deterministic` only.

Reply:
Seeing the failed results there is another obvious issue with allowing foldable inferred filters - the idempotence is broken. I would very much like to keep this rule idempotent. 

Should I try to include foldable inferred filters while keeping it idempotent or shall I revert the last change?

Reply:
Ah, ok. If so, its okay to revert it. Could you leave some comments about why we need the foldable check there?